Authentikos: Refactor Token/Secret creators to interfaces.

wlynch opened the pull request:
This allows for other token types to be created while reusing the same
common creation / reconcilation components.

As an example for how this can be extended, adds a GitHub App token
creator as an alternative to the existing Google OAuth implementation.

Also changes secret updates to patches of just the data to fix an issue
dicovered during testing where the reconciler would overwrite any
user-added labels/annotations.

Implements https://github.com/istio/test-infra/issues/2659

wlynch committed:
Authentikos: Refactor Token/Secret creators to interfaces.

This allows for other token types to be created while reusing the same
common creation / reconcilation components.

As an example for how this can be extended, adds a GitHub App token
creator as an alternative to the existing Google OAuth implementation.

Also changes secret updates to patches of just the data to fix an issue
dicovered during testing where the reconciler would overwrite any
user-added labels/annotations.

istio-testing commented:
Hi @wlynch. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Ftest-infra).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

wlynch commented on repos.bzl:
```diff
@@ -2,944 +2,58 @@ load("@bazel_gazelle//:deps.bzl", "go_repository")
 
 def go_repositories():
```
Bit of a large diff here, though looks like a lot of reorderings - this was generated via `gazelle update-repos -from_file=go.mod -to_macro=repos.bzl%go_repositories` from the authentikos directory. Let me know if there's a better way to update this file!

ericvn commented:
/ok-to-test

istio-testing commented:
@wlynch: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-test-authentikos_test-infra | 9e621722d26901dabe76c2fb0cad0558c8d6bff5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1323387660311465984) | `/test unit-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-test-authentikos_test-infra | 9e621722d26901dabe76c2fb0cad0558c8d6bff5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1323387660311465984) | `/test unit-test-authentikos_test-infra`
lint_test-infra | 9e621722d26901dabe76c2fb0cad0558c8d6bff5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/lint_test-infra/1323387660214996992) | `/test lint_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-test-authentikos_test-infra | 9e621722d26901dabe76c2fb0cad0558c8d6bff5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1323387660311465984) | `/test unit-test-authentikos_test-infra`
lint_test-infra | 9e621722d26901dabe76c2fb0cad0558c8d6bff5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/lint_test-infra/1323387660214996992) | `/test lint_test-infra`
integ-test-authentikos_test-infra | 9e621722d26901dabe76c2fb0cad0558c8d6bff5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1323387660357603328) | `/test integ-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

wlynch committed:
Authentikos: Add missing license headers.

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_test-infra | 9e621722d26901dabe76c2fb0cad0558c8d6bff5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/lint_test-infra/1323387660214996992) | `/test lint_test-infra`
integ-test-authentikos_test-infra | 9e621722d26901dabe76c2fb0cad0558c8d6bff5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1323387660357603328) | `/test integ-test-authentikos_test-infra`
unit-test-authentikos_test-infra | 4e5e897a6b5d27a87d4943db9ac64bc37869c567 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1324039296260247552) | `/test unit-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-test-authentikos_test-infra | 4e5e897a6b5d27a87d4943db9ac64bc37869c567 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1324039296260247552) | `/test unit-test-authentikos_test-infra`
integ-test-authentikos_test-infra | 4e5e897a6b5d27a87d4943db9ac64bc37869c567 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1324039296281219072) | `/test integ-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

wlynch committed:
Authentikos: Fix gitignore to not ignore dirs with "authentikos" in
path.

The gitignore was configured to ignore the authentikos binary generated
by `go build`, but because of the gitignore syntax this resulted in git
ignoring any directory with "authentikos" in the path.

Also adds in missing files that were previously being ignored.

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-test-authentikos_test-infra | 4e5e897a6b5d27a87d4943db9ac64bc37869c567 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1324039296281219072) | `/test integ-test-authentikos_test-infra`
unit-test-authentikos_test-infra | 248249b83e9715fcb86053d495e203475498e051 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1324047812479946752) | `/test unit-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-test-authentikos_test-infra | 4e5e897a6b5d27a87d4943db9ac64bc37869c567 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1324039296281219072) | `/test integ-test-authentikos_test-infra`
unit-test-authentikos_test-infra | 1629aa31b54d1e1a13b748c88ab0a2cdf2393e15 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1324048620260954112) | `/test unit-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-test-authentikos_test-infra | 1629aa31b54d1e1a13b748c88ab0a2cdf2393e15 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1324048620260954112) | `/test unit-test-authentikos_test-infra`
integ-test-authentikos_test-infra | 1629aa31b54d1e1a13b748c88ab0a2cdf2393e15 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1324048620361617408) | `/test integ-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-test-authentikos_test-infra | 1629aa31b54d1e1a13b748c88ab0a2cdf2393e15 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1324048620361617408) | `/test integ-test-authentikos_test-infra`
unit-test-authentikos_test-infra | a8a6fd7dcb834c70544dffe5aaed850f5a4bcfb9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1324057401162207232) | `/test unit-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@wlynch: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-test-authentikos_test-infra | a8a6fd7dcb834c70544dffe5aaed850f5a4bcfb9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/unit-test-authentikos_test-infra/1324057401162207232) | `/test unit-test-authentikos_test-infra`
integ-test-authentikos_test-infra | a8a6fd7dcb834c70544dffe5aaed850f5a4bcfb9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1324057401183178752) | `/test integ-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

wlynch committed:
Authentikos: Fix validation test when running as root.

If running as root (i.e. default user for docker run), setting the file
permissions to `0000` doesn't actually prevent the file from being
unreadable. This was causing false signals in CI testing. This modifies
the behavior to just create a file that doesn't exist, which has the
same effect of what we're trying to test (e.g. that we can't read the
template file).

Since the deletedCredsFile is similar in this sense this was also
changed to be consistent.

Also added some minor test QoL changes (usage of t.Helper,
t.Log) to aid future debugging.

istio-testing commented:
@wlynch: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-test-authentikos_test-infra | ea61aa0ec3a73b7c5da2043c5acbe6dd6c659e5a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/integ-test-authentikos_test-infra/1324080381934178304) | `/test integ-test-authentikos_test-infra`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

wlynch committed:
Authentikos: Fix klog and pflag flag parsing.

Adds missing support for mixing standard golang and pflag flags. See
https://github.com/spf13/pflag#supporting-go-flags-when-using-pflag for
details.

wlynch committed:
Authentikos: Fix integration tests.

This includes fixes for a variety of issues related to integration
testing:

- Use `kind load docker-image` to pass in the local image to the test.
  - Disable remote pulling of authentikos images to prevent pulling a
    non-test image
  - Modify integ-test makefile rule to depend on fresh image builds.
- Update `kindest/node` version to latest v1.17.x image. This is
  required to support host networking DNS resolution that is broken in
  v1.17.0.
- Modify unit tests to be more OSX friendly (`xargs -r` doesn't exist on
  OSX)

wlynch committed:
Authentikos: Fix log formatting commands.

istio-testing commented:
@wlynch: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@wlynch: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pull-test-infra-check-testgrid-config-transfigure | 48fde298aae1835ad04d9eccf14c376c9a7a8fc1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_test-infra/3036/pull-test-infra-check-testgrid-config-transfigure/1433120901024452608) | `/test pull-test-infra-check-testgrid-config-transfigure`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
